app/abci/utils.go

```diff
@@ -0,0 +1,232 @@
+// vote extensions are enabled at height h, then a proposer will receive vote extensions
+// in height h+1. This is primarily utilized by any module that needs to make state changes
+// based on whether they were included in a proposal.
+func VoteExtensionsEnabled(ctx sdk.Context) bool {
+	cp := ctx.ConsensusParams()
+	if cp.Abci == nil || cp.Abci.VoteExtensionsEnableHeight == 0 {
+		return false
+	}
+
+	// Per the cosmos sdk, the first block should not utilize the latest finalize block state. This means
+	// vote extensions should NOT be making state changes.
+	//
+	// Ref: https://github.com/cosmos/cosmos-sdk/blob/2100a73dcea634ce914977dbddb4991a020ee345/baseapp/baseapp.go#L488-L495
+	if ctx.BlockHeight() <= 1 {
+		return false
+	}
+
+	return cp.Abci.VoteExtensionsEnableHeight < ctx.BlockHeight()
+}
+
+// ValidatorStore defines the interface contract require for verifying vote
+// extension signatures. Typically, this will be implemented by the x/staking
+// module, which has knowledge of the CometBFT public key.
+type ValidatorStore interface {
+	GetPubKeyByConsAddr(context.Context, sdk.ConsAddress) (cmtprotocrypto.PublicKey, error)
+}
+
+// ValidateVoteExtensions defines a helper function for verifying vote extension
+// signatures that may be passed or manually injected into a block proposal from
+// a proposer in PrepareProposal. It returns an error if any signature is invalid
+// or if unexpected vote extensions and/or signatures are found or less than 2/3
+// power is received.
+func ValidateVoteExtensions(
+	ctx sdk.Context,
+	valStore ValidatorStore,
+	extCommit cometabci.ExtendedCommitInfo,
+) error {
+	currentHeight := ctx.HeaderInfo().Height
+	chainID := ctx.HeaderInfo().ChainID
+	commitInfo := ctx.CometInfo().GetLastCommit()
+
+	// Check that both extCommit + commit are ordered in accordance with vp/address.
+	if err := ValidateExtendedCommitAgainstLastCommit(extCommit, commitInfo); err != nil {
+		return err
+	}
+
+	// Start checking vote extensions only **after** the vote extensions enable
+	// height, because when `currentHeight == VoteExtensionsEnableHeight`
+	// PrepareProposal doesn't get any vote extensions in its request.
+	extensionsEnabled := VoteExtensionsEnabled(ctx)
+	marshalDelimitedFn := func(msg proto.Message) ([]byte, error) {
+		var buf bytes.Buffer
+		if err := protoio.NewDelimitedWriter(&buf).WriteMsg(msg); err != nil {
+			return nil, err
+		}
+
+		return buf.Bytes(), nil
+	}
+
+	var (
+		// Total voting power of all vote extensions.
+		totalVP int64
+		// Total voting power of all validators that submitted valid vote extensions.
+		sumVP int64
+	)
+
+	for _, vote := range extCommit.Votes {
+		totalVP += vote.Validator.Power
+
+		if extensionsEnabled {
+			if vote.BlockIdFlag == cmtproto.BlockIDFlagCommit && len(vote.ExtensionSignature) == 0 {
+				return fmt.Errorf("vote extension signature is missing; validator addr %s",
+					vote.Validator.String(),
+				)
+			}
```

Comment:
Could this be an issue for a new active validator who doesn't sign the first batch?